state-chain/custom-rpc/src/pool_client.rs

```diff
@@ -0,0 +1,610 @@
+	ExtrinsicNotFound(usize),
+	#[error("Failed to submit extrinsic to the transaction pool after {0} attempts")]
+	PoolSubmitError(usize),
+	#[error("Could not acquire lock for transaction pool")]
+	PoolLockingError,
+	#[error("{0:?}")]
+	TransactionStatusError(&'static str),
+
+	#[error("{0:?}")]
+	CodecError(#[from] codec::Error),
+	#[error("{0:?}")]
+	RuntimeApiError(#[from] ApiError),
+	#[error("{0:?}")]
+	SubstrateClientError(#[from] sc_client_api::blockchain::Error),
+	#[error("{0:?}")]
+	TransactionPoolError(#[from] sc_transaction_pool::error::Error),
+	#[error("{0:?}")]
+	TransactionValidityError(#[from] TransactionValidityError),
+	#[error("{0:?}")]
+	ExtrinsicDispatchError(#[from] error_decoder::DispatchError),
+	#[error("{0:?}")]
+	ExtrinsicDynamicEventsError(#[from] events_decoder::DynamicEventError),
+	#[error(transparent)]
+	ErrorObject(#[from] ErrorObjectOwned),
+}
+
+pub struct SignedPoolClient<C, B, BE>
+where
+	B: BlockT<Hash = state_chain_runtime::Hash, Header = state_chain_runtime::Header>,
+	BE: Send + Sync + 'static + Backend<B>,
+	C: Send
+		+ Sync
+		+ 'static
+		+ BlockBackend<B>
+		+ HeaderBackend<B>
+		+ HeaderMetadata<B, Error = sc_client_api::blockchain::Error>
+		+ CallApiAt<B>
+		+ StorageProvider<B, BE>
+		+ sp_api::ProvideRuntimeApi<B>
+		+ sp_runtime::traits::BlockIdTo<B>,
+	C::Api: CustomRuntimeApi<B>
+		+ sp_api::Core<B>
+		+ sp_api::Metadata<B>
+		+ sp_block_builder::BlockBuilder<B>
+		+ sp_transaction_pool::runtime_api::TaggedTransactionQueue<B>
+		+ frame_system_rpc_runtime_api::AccountNonceApi<B, AccountId, Nonce>,
+{
+	client: Arc<C>,
+	pool: Arc<FullPool<B, C>>,
+	pool_semaphore: Arc<Semaphore>,
+	_phantom: PhantomData<(B, BE)>,
+	system_api: System<FullPool<B, C>, C, B>,
+	pair_signer: PairSigner<sp_core::sr25519::Pair>,
+	runtime_decoders: Arc<RwLock<HashMap<u32, RuntimeDecoder>>>,
+}
+
+impl<C, B, BE> SignedPoolClient<C, B, BE>
+where
+	B: BlockT<Hash = state_chain_runtime::Hash, Header = state_chain_runtime::Header>,
+	BE: Send + Sync + 'static + Backend<B>,
+	C: Send
+		+ Sync
+		+ 'static
+		+ BlockBackend<B>
+		+ HeaderBackend<B>
+		+ HeaderMetadata<B, Error = sc_client_api::blockchain::Error>
+		+ CallApiAt<B>
+		+ StorageProvider<B, BE>
+		+ sp_api::ProvideRuntimeApi<B>
+		+ sp_runtime::traits::BlockIdTo<B>,
+	C::Api: CustomRuntimeApi<B>
+		+ sp_api::Core<B>
+		+ sp_api::Metadata<B>
+		+ sp_block_builder::BlockBuilder<B>
+		+ sp_transaction_pool::runtime_api::TaggedTransactionQueue<B>
+		+ frame_system_rpc_runtime_api::AccountNonceApi<B, AccountId, Nonce>,
+{
+	pub fn new(client: Arc<C>, pool: Arc<FullPool<B, C>>, pair: sp_core::sr25519::Pair) -> Self {
+		Self {
+			_phantom: Default::default(),
+			system_api: System::new(client.clone(), pool.clone(), sc_rpc_api::DenyUnsafe::Yes),
+			pair_signer: PairSigner::new(pair),
+			runtime_decoders: Arc::new(RwLock::new(HashMap::new())),
+			pool_semaphore: Arc::new(Semaphore::new(1)), // Allow only 1 pool access at a time
```

Comment:
How is this is different from a mutex? (Or what's the benefit of using a semaphore over a mutex for this?)